### PR TITLE
feat(trends): Rework formula experience to "formula mode"

### LINKED ISSUE #12069: Let the number visualisation of the trend insights visualise formula results
## Is your feature request related to a problem?

I want to calculate a Conversion Rate for the website and visualize this.
I have created the CR by using the trends report and taking two series (A and B) with a formula (B/A)

## Describe the solution you'd like

To have the number visualization show the result of the formula.

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

There are two problems with formulas:
1. The relationship between insight series and the formula has been a bit confusing. It's not _super_ clear that the formula becomes the singular output of the insight, because it's tacked on as an "advanced option".
2. It wasn't possible to use formulas with the "Number" chart type because that one only supports a single output, and there wasn't logic in place to determine that a formula results in a single output. A bunch of users have requested this.

## Changes

This reworks the formula experience to "formula mode", in which "Series" are instead named "Variables", and the formula input takes over the spotlight:

Additionally, the revamped logic allows for formula-based Number insights:

Screen Shot 2022-10-07 at 16 35 58

Closes #12069.

## How did you test this code?

Lots of testing edge cases manually.

### CODE REVIEW

**Comment by timgl** [comment]
Nice work! I think this could probably use a little bit of design love, i'm not a huge fan of that switch. I think we can probably do something a bit nicer, like some 'add formula' text that turns into an input or something? @corywatilo

**Comment by Twixes** [comment]
Alternative idea:
Formula mode

So, I'm not sure if the existing "Add formula" is the right way to phrase this, because the way an insight with a formula works is fundamentally different from one without a formula (series become variables). So it's more like a "mode" instead of an "add-on".
An additional consideration is that with series being defined on the left side of the query builder, it's most obvious if the formula is displayed there too (as it takes over _all_ the series), but we probably should not give it _too_ much visual weight – most insights don't rely on a formula.
I'm definitely up for considering other UI solutions in any case.

**Comment by annikaschmid** [comment]
I like that the formula is now under the Series / Variables, it makes more sense having it closer to those than to the breakdown.

I prefer the UI in the newer screenshot, where we don't have a toggle and display the field right from the start. That way, it's immediately obvious what I can do. @Twixes is there a reason you removed the tool tip (i) which we have in the current UI?

**Comment by mariusandra** [comment]
Could we support multiple formulas? This is something I've heard asked before.

image

- Clicking "add formula" would add line "C" where you can write "A + B"
- Clicking "add formula" once more would open line "D" where you can write "(A - B) / C"
- Clicking "add graph series" would add another trend line "E".

**Comment by Twixes** [comment]
All possible but more complex… @mariusandra The UX is the trickiest part IMO. We can definitely consider this for some sprint, though wouldn't for this one.

@annikaschmid The existing tooltip didn't seem particularly valuable, it mostly just explained what format the expression could be in, but I think that's more naturally presented in the input placeholder. I agree some intro tooltip can be useful though, so added this:
Screen Shot 2022-10-10 at 18 52 31

File: frontend/src/scenes/insights/EditorFilters/TrendsFormula.tsx [hunk 1]
```diff
@@ -1,74 +1,69 @@
...
+export function TrendsFormula({ filters, insightProps }: EditorFilterProps): JSX.Element | null {
+    const { isFormulaOn } = useValues(trendsLogic(insightProps))
     const { setFilters } = useActions(trendsLogic(insightProps))
-    const [isUsingFormulas, setIsUsingFormulas] = useState(!!filters.formula)
-    const formulaEnabled = (filters.events?.length || 0) + (filters.actions?.length || 0) > 0
     const [value, setValue] = useState(filters.formula)
 
     useEffect(() => {
-        setValue(filters.formula)
+        // Don't clear the formula so that the value is still there after toggling the formula switch
```
Comment (pauldambra) [inline frontend/src/scenes/insights/EditorFilters/TrendsFormula.tsx:16]: 💖

**Review by pauldambra — APPROVED** [review]
LGTM

Way better placement to be closely associated with series

**Review by mariusandra — COMMENTED** [review]
Looks good and definitely an improvement! The only thing that didn't work for me was this button:

**Comment by Twixes** [comment]
Yeah, that's a separate issue: #9478 @mariusandra To be fixed separately.

**Pull request merged**